src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala

```diff
@@ -180,6 +182,65 @@ class FastSyncItSpec extends FlatSpecBase with Matchers with BeforeAndAfterAll {
     }
   }
 
+  it should "switch to regular sync once `safeDownloadTarget` is reached" in customTestCaseResourceM(
+    FakePeer.start3FakePeersRes()
+  ) { case (peer1, peer2, peer3) =>
+    for {
+      _ <- peer2.importBlocksUntil(1200)(IdentityUpdate)
+      _ <- peer3.importBlocksUntil(1200)(IdentityUpdate)
+      _ <- peer1.connectToPeers(Set(peer2.node, peer3.node))
+      _ <- peer1.startFastSync().delayExecution(50.milliseconds)
+      _ <- peer1.waitForFastSyncFinish()
+    } yield {
+      assert(peer1.bl.getBestBlockNumber() == peer3.bl.getBestBlockNumber() - peer3.testSyncConfig.pivotBlockOffset)
+    }
+  }
+
+  it should "blacklist peer on Invalid batch last header number" in customTestCaseResourceM(
+    FakePeer.start3FakePeersRes()
+  ) { case (peer1, peer2, peer3) =>
+    for {
+      _ <- peer2.importBlocksUntil(1000)(IdentityUpdate)
+      _ <- peer3.importMaliciousBlockNumbers(201, 1200)(IdentityUpdate)
+
+      _ <- peer1.connectToPeers(Set(peer2.node, peer3.node))
+      _ <- peer1.startFastSync().delayExecution(50.milliseconds)
+      _ <- peer1.waitForFastSyncFinish()
+    } yield {
+      // Peer3 is blacklisted
+      val blacklistedPeer = PeerId(s"${peer3.node.addr.getHostAddress}:${peer3.node.tcpPort}")
+      val blacklistReason = peer1.blacklist.cache.getIfPresent(blacklistedPeer)
+
+      assert(peer1.blacklist.isBlacklisted(blacklistedPeer))
+      assert(blacklistReason.get == BlacklistReasonType.BlockHeaderValidationFailedType)
+    }
+  }
+
+  it should "start branch resolving on invalid batch last header hash" in customTestCaseResourceM(
```

Comment:
should the test scenario be renamed? Or how do we check that  branch resolving happened?

Reply:
The occurrence of branch resolving can only be identified from logs. Maybe renaming the test is better.